File: src/server/src/authoring/api/authoring.controller.ts
import {
  BadRequestException,
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Logger,
  Post,
  ServiceUnavailableException,
} from '@nestjs/common';
import {
  ApiBearerAuth,
  ApiOperation,
  ApiResponse,
  ApiTags,
} from '@nestjs/swagger';
import { Roles } from '../../auth/decorators/roles.decorator';
import { Role } from '../../auth/roles.enum';
import { VocabularyProjectionService } from '../../rules/vocabulary-projection.service';
import { parseRuleJson } from '../../vdf/api/rule-json.helper';
import { AuthoringService } from '../authoring.service';
import { RuleInterpreterService } from '../llm/rule-interpreter.service';
import {
  DryRunResponseDto,
  InterpretRequestDto,
  InterpretResponseDto,
  LintReportDto,
  ParaphraseResponseDto,
  RuleJsonRequestDto,
  VocabularyResponseDto,
} from './authoring.dto';

/**
 * Authoring tools: vocabulary tree, natural-language interpretation, registry-grounded
 * linting, deterministic paraphrasing, and corpus dry-run preview. These are read-only
 * / non-persisting — they help an author shape a rule before it is saved via the rules
 * controller.
 *
 * The vocabulary tree is readable by any authenticated principal (the scope-picker
 * source). The mutating authoring actions require the Author role; per-action @Roles
 * keeps the vocabulary read open while gating the rest.
 */
@ApiTags('authoring')
@ApiBearerAuth()
@Controller('api/authoring')
export class AuthoringController {
  private readonly logger = new Logger(AuthoringController.name);

  constructor(
    private readonly authoring: AuthoringService,
    private readonly interpreter: RuleInterpreterService,
    private readonly vocabulary: VocabularyProjectionService,
  ) {}

  @Get('vocabulary')
  @ApiOperation({
    summary:
      'The controlled vocabulary as an object -> property tree (Active only), plus ' +
      'operator and outcome names. The authoring scope-picker source.',
  })
  @ApiResponse({ status: 200, type: VocabularyResponseDto })
  async vocabularyTree(): Promise<VocabularyResponseDto> {
    const tree = await this.vocabulary.projectTree();
    return VocabularyResponseDto.from(tree);
  }

  @Post('interpret')
  @Roles(Role.Author)
  @HttpCode(HttpStatus.OK)
  @ApiOperation({
    summary:
      'Interpret a natural-language rule into a candidate rule definition, grounded ' +
      'on the (optionally scoped) registry vocabulary.',
  })
  @ApiResponse({ status: 200, type: InterpretResponseDto })
  @ApiResponse({ status: 400, description: 'Unknown vocabulary scope.' })
  @ApiResponse({ status: 503, description: 'The interpreter is unavailable.' })
  async interpret(
    @Body() request: InterpretRequestDto,
  ): Promise<InterpretResponseDto> {
    // Resolve the SCOPED grounding subjects from the registry projection. An unknown
    // object/property is a 400 (the UI can never silently scope to nothing) — mirrors
    // the .NET scoped-interpret.
    const scope = await this.vocabulary.resolveScope(
      request.objects,
      request.properties,
    );
    if (!scope.ok) {
      throw new BadRequestException(scope.error ?? 'Unknown vocabulary scope.');
    }

    try {
      const result = await this.interpreter.interpretScoped(
        request.naturalLanguage,
        scope.subjects,
      );
      return InterpretResponseDto.from(result);
    } catch (error) {
      // The facade falls back to the offline stub on a live failure, so reaching here
      // means the grounding itself was unavailable. Degrade to 503 — never a 500 that
      // could leak provider/config detail.
      const message = error instanceof Error ? error.message : String(error);
      this.logger.warn(`Interpreter unavailable: ${message}`);
      throw new ServiceUnavailableException(
        'The rule interpreter is currently unavailable. Contact your administrator if this persists.',
      );
    }
  }

  @Post('lint')
  @Roles(Role.Author)
  @HttpCode(HttpStatus.OK)
  @ApiOperation({
    summary: 'Lint a rule JSON object against the live registry vocabulary.',
  })
  @ApiResponse({ status: 200, type: LintReportDto })
  async lint(@Body() request: RuleJsonRequestDto): Promise<LintReportDto> {
    const rule = parseRuleJson(request.ruleJson);
    const report = await this.authoring.lint(rule);
    return LintReportDto.from(report);
  }

  @Post('paraphrase')
  @Roles(Role.Author)
  @HttpCode(HttpStatus.OK)
  @ApiOperation({
    summary: 'Produce a deterministic English paraphrase of a rule.',
  })
  @ApiResponse({ status: 200, type: ParaphraseResponseDto })
  paraphrase(@Body() request: RuleJsonRequestDto): ParaphraseResponseDto {
    const rule = parseRuleJson(request.ruleJson);
    return { paraphrase: this.authoring.paraphrase(rule) };
  }

  @Post('dry-run')
  @Roles(Role.Author)
  @HttpCode(HttpStatus.OK)
  @ApiOperation({
    summary:
      'Dry-run a candidate rule against the repo fixtures corpus (no side effects).',
  })
  @ApiResponse({ status: 200, type: DryRunResponseDto })
  async dryRun(
    @Body() request: RuleJsonRequestDto,
  ): Promise<DryRunResponseDto> {
    const rule = parseRuleJson(request.ruleJson);
    const result = await this.authoring.previewFromRepoFixtures(rule);
    return DryRunResponseDto.from(result);
  }
}

File: src/backend/IAW.Vdf.Api/Controllers/AuthoringController.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using IAW.Vdf.Api.Auth;
using IAW.Vdf.Api.Dtos;
using IAW.Vdf.Api.Infrastructure;
using IAW.Vdf.Abstractions.Authoring;
using IAW.Vdf.Abstractions.Vocabulary;
using IAW.Vdf.Authoring.DryRun;
using IAW.Vdf.Authoring.Linting;
using IAW.Vdf.Authoring.Paraphrase;
using IAW.Vdf.Core.Serialization;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace IAW.Vdf.Api.Controllers;

/// <summary>
/// Authoring tools (Author role): natural-language interpretation, vocabulary linting, round-trip
/// paraphrasing, and corpus dry-run preview. These are read-only / non-persisting — they help an author
/// shape a rule before it is saved via <c>RulesController</c>.
/// </summary>
[ApiController]
[Route("api/authoring")]
[Authorize(Policy = VdfPolicies.CanAuthor)]
public sealed class AuthoringController : ControllerBase
{
    private readonly IRuleInterpreter _interpreter;
    private readonly VocabularyCatalog _vocabulary;
    private readonly VocabularyLinter _linter;
    private readonly RoundTripParaphraser _paraphraser;
    private readonly DryRunPreviewer _previewer;
    private readonly ILogger<AuthoringController> _logger;

    /// <summary>Creates the controller.</summary>
    public AuthoringController(
        IRuleInterpreter interpreter,
        VocabularyCatalog vocabulary,
        VocabularyLinter linter,
        RoundTripParaphraser paraphraser,
        DryRunPreviewer previewer,
        ILogger<AuthoringController> logger)
    {
        _interpreter = interpreter;
        _vocabulary = vocabulary;
        _linter = linter;
        _paraphraser = paraphraser;
        _previewer = previewer;
        _logger = logger;
    }

    /// <summary>Interprets a natural-language rule into a candidate rule definition.</summary>
    /// <param name="request">The natural-language request.</param>
    /// <param name="cancellationToken">A cancellation token.</param>
    /// <returns>The candidate, confidence, unmapped phrases, and gaps; 503 when the interpreter is unavailable.</returns>
    [HttpPost("interpret")]
    [ProducesResponseType(typeof(InterpretResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status503ServiceUnavailable)]
    public async Task<ActionResult<InterpretResponse>> Interpret(
        [FromBody] InterpretRequest request,
        CancellationToken cancellationToken)
    {
        InterpretationResult result;
        try
        {
            result = await _interpreter
                .InterpretAsync(request.NaturalLanguage, _vocabulary, cancellationToken)
                .ConfigureAwait(false);
        }
        catch (InvalidOperationException ex)
        {
            // Interpreter disabled / no key configured — degrade gracefully, never 500.
            // M1: never leak ex.Message to the client (it can disclose internal config); the full
            // message is logged server-side below for operators.
            _logger.LogWarning("Interpreter unavailable: {Message}", ex.Message);
            return Problem(
                title: "The rule interpreter is currently unavailable.",
                detail: "The rule interpreter is currently unavailable. Contact your administrator if this persists.",
                statusCode: StatusCodes.Status503ServiceUnavailable);
        }

        JsonElement? candidate = null;
        if (result.Candidate is not null)
        {
            using var doc = JsonDocument.Parse(RuleSerializer.Serialize(result.Candidate));
            candidate = doc.RootElement.Clone();
        }

        return Ok(new InterpretResponse
        {
            Candidate = candidate,
            Confidence = result.Confidence,
            UnmappedPhrases = result.UnmappedPhrases,
            Gaps = result.Gaps,
        });
    }

    /// <summary>Lints a rule JSON object against the vocabulary and reference data.</summary>
    /// <param name="request">The rule JSON request.</param>
    /// <returns>The lint report.</returns>
    [HttpPost("lint")]
    [ProducesResponseType(typeof(LintReportDto), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public ActionResult<LintReportDto> Lint([FromBody] RuleJsonRequest request)
    {
        if (request.RuleJson.ValueKind != JsonValueKind.Object)
        {
            return Problem(title: "ruleJson must be a JSON object.", statusCode: StatusCodes.Status400BadRequest);
        }

        var report = _linter.LintJson(request.RuleJson.GetRawText());
        return Ok(LintReportDto.From(report));
    }

    /// <summary>Produces a deterministic English paraphrase of a rule.</summary>
    /// <param name="request">The rule JSON request.</param>
    /// <returns>The paraphrase.</returns>
    [HttpPost("paraphrase")]
    [ProducesResponseType(typeof(ParaphraseResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public ActionResult<ParaphraseResponse> Paraphrase([FromBody] RuleJsonRequest request)
    {
        if (!RuleJsonHelper.TryParse(request.RuleJson, out var rule, out var error))
        {
            return Problem(title: error, statusCode: StatusCodes.Status400BadRequest);
        }

        var text = _paraphraser.Paraphrase(rule!);
        return Ok(new ParaphraseResponse { Paraphrase = text });
    }

    /// <summary>Dry-runs a candidate rule against the repository fixtures corpus (no side effects).</summary>
    /// <param name="request">The rule JSON request.</param>
    /// <param name="cancellationToken">A cancellation token.</param>
    /// <returns>The per-fixture dry-run hits.</returns>
    [HttpPost("dry-run")]
    [ProducesResponseType(typeof(DryRunResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<ActionResult<DryRunResponse>> DryRun(
        [FromBody] RuleJsonRequest request,
        CancellationToken cancellationToken)
    {
        if (!RuleJsonHelper.TryParse(request.RuleJson, out var rule, out var error))
        {
            return Problem(title: error, statusCode: StatusCodes.Status400BadRequest);
        }

        var result = await _previewer
            .PreviewFromRepoFixturesAsync(rule!, cancellationToken)
            .ConfigureAwait(false);

        return Ok(DryRunResponse.From(result));
    }
}
